File: python/open_guardrail/guards/geographic_restrict.py
"""Geographic restriction guard for compliance."""
from __future__ import annotations

import re
import time
from typing import List

from open_guardrail.core import GuardResult


class _GeographicRestrict:
    def __init__(
        self,
        *,
        action: str = "block",
        restricted_regions: List[str],
    ) -> None:
        self.name = "geographic-restrict"
        self.action = action
        self._regions = restricted_regions
        self._patterns = [
            re.compile(
                r"\b" + re.escape(r) + r"\b",
                re.IGNORECASE,
            )
            for r in restricted_regions
        ]

    def check(
        self, text: str, stage: str = "input"
    ) -> GuardResult:
        start = time.perf_counter()
        matched: list[str] = []

        for i, pat in enumerate(self._patterns):
            if pat.search(text):
                matched.append(self._regions[i])

        triggered = len(matched) > 0
        score = min(len(matched) / 2, 1.0) if triggered else 0.0
        elapsed = (time.perf_counter() - start) * 1000

        return GuardResult(
            guard_name="geographic-restrict",
            passed=not triggered,
            action=self.action if triggered else "allow",
            score=score,
            message=(
                "Restricted region mentioned"
                if triggered
                else None
            ),
            latency_ms=round(elapsed, 2),
            details=(
                {"matched_regions": matched}
                if triggered
                else None
            ),
        )


def geographic_restrict(
    *,
    action: str = "block",
    restricted_regions: List[str],
) -> _GeographicRestrict:
    return _GeographicRestrict(
        action=action,
        restricted_regions=restricted_regions,
    )

File: packages/guards/src/geographic-restrict.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface GeographicRestrictOptions {
  action: 'block' | 'warn';
  restrictedRegions: string[];
}

export function geographicRestrict(options: GeographicRestrictOptions): Guard {
  const patterns = options.restrictedRegions.map(
    (region) => new RegExp(`\\b${region.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')}\\b`, 'i'),
  );

  return {
    name: 'geographic-restrict',
    version: '0.1.0',
    description: 'Geographic restriction guard for compliance',
    category: 'compliance',
    supportedStages: ['input'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const matchedRegions: string[] = [];

      for (let i = 0; i < patterns.length; i++) {
        if (patterns[i].test(text)) {
          matchedRegions.push(options.restrictedRegions[i]);
        }
      }

      const triggered = matchedRegions.length > 0;
      const score = triggered ? Math.min(matchedRegions.length / 2, 1.0) : 0;

      return {
        guardName: 'geographic-restrict',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        score,
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { matchedRegions } : undefined,
      };
    },
  };
}
